tmt/steps/prepare/feature.py

```diff
@@ -0,0 +1,197 @@
+import re
+from typing import List, Optional
+
+import tmt
+import tmt.base
+import tmt.log
+import tmt.options
+import tmt.steps
+import tmt.steps.prepare
+import tmt.utils
+from tmt.steps.provision import Guest
+from tmt.utils import Command, field
+
+
+class Feature(tmt.utils.Common):
+    """ Base class for feature implementations """
+
+    KEY: str
+
+    def __init__(
+            self,
+            *,
+            parent: tmt.steps.prepare.PreparePlugin,
+            guest: Guest,
+            logger: tmt.log.Logger) -> None:
+        """ Initialize feature data """
+        super().__init__(logger=logger, parent=parent, relative_indent=0)
+        self.guest = guest
+
+    def get_guest_distro(self, guest: Guest, logger: tmt.log.Logger) -> Optional[str]:
+        """ Get guest distro by parsing the guest facts """
+        distro = self.guest.facts.distro
+        if distro is None:
+            return None
+
+        if re.search(r'Fedora Linux \d+', distro):
+            return 'Fedora'
+        if re.search(r'CentOS Linux.*7\.', distro):
+            return 'CentOS-7'
+        if re.search(r'CentOS Stream.*8', distro):
+            return 'CentOS-Stream-8'
+        if re.search(r'CentOS Stream.*9', distro):
+            return 'CentOS-Stream-9'
+        if re.search(r'Red Hat Enterprise Linux.*8\.', distro):
+            return 'RHEL-8'
+        if re.search(r'Red Hat Enterprise Linux.*9\.', distro):
+            return 'RHEL-9'
+        return None
+
+
+class ToggleableFeature(Feature):
+    def enable(self, guest: Guest, logger: tmt.log.Logger) -> None:
+        raise NotImplementedError
+
+    def disable(self, guest: Guest, logger: tmt.log.Logger) -> None:
+        raise NotImplementedError
+
+
+class CRB(ToggleableFeature):
+    KEY = 'crb'
+    # TBD
+
+
+class FIPS(ToggleableFeature):
+    KEY = 'fips'
+    # TBD
+
+
+class EPEL(ToggleableFeature):
+    KEY = 'epel'
+
+    def enable(self, guest: Guest, logger: tmt.log.Logger) -> None:
+        guest_distro = self.get_guest_distro(guest=guest, logger=logger)
+        if guest_distro is None:
+            raise tmt.utils.PrepareError('The distro of the guest is not supported.')
+
+        if guest_distro == 'Fedora':
+            self.info('Enable EPEL on Fedora, do nothing ...')
+        elif guest_distro == 'CentOS-Stream-8':
+            # dnf config-manager --set-enabled powertools
+            # dnf install epel-release epel-next-release
+            self.info('Enable EPEL on CentOS Stream 8')
+            command = Command()
+            if self.guest.facts.is_superuser is False:
+                command += Command('sudo')
+            command1 = command + Command('dnf', 'config-manager', '--set-enabled', 'powertools')
+            command2 = command + Command('dnf', '-y', 'install',
+                                         'epel-release', 'epel-next-release')
+            self.guest.execute(command1, silent=True)
+            self.guest.execute(command2, silent=True)
```

Comment:
Instead of vague `command1` and `command2`, maybe we could use `ShellScript`:

```python
sudo = 'sudo' if self.guest.facts.is_superuser is False else ''

self.guest.execute(
      ShellScript(f'{sudo} dnf config-manager --set-enabled powertools') 
  and ShellScript(f'{sudo} dnf -y install epel-release epel-next-release'),
  silent=True)
```

More verbose though, but it seems more readable to me. What do you think?

Reply:
Hi @happz, looks the `ShellScript` before **and** is not executed on the guest. e.g.

```python
> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/steps/prepare/feature.py(101)enable()
-> if guest_distro == Distro.FEDORA:
(Pdb) l
 96  	        if guest_distro is None:
 97  	            raise tmt.utils.PrepareError('The distro of the guest is not supported.')
 98  	
 99  	        import pdb; pdb.set_trace()
100  	
101  ->	        if guest_distro == Distro.FEDORA:
102  	            self.info('Enable EPEL on Fedora, do nothing ...')
103  	            sudo = 'sudo' if self.guest.facts.is_superuser is False else ''
104  	            cmd1 = 'echo subscription-manager repos --enable codeready-builder-for-rhel-8-$(arch)-rpms'
105  	            cmd2 = 'echo dnf -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm'
106  	            self.guest.execute(ShellScript(f'{sudo} {cmd1}') and ShellScript(f'{sudo} {cmd2}'),
(Pdb) b 106
Breakpoint 1 at /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/steps/prepare/feature.py:106
(Pdb) c
        Enable EPEL on Fedora, do nothing ...
> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/steps/prepare/feature.py(106)enable()
-> self.guest.execute(ShellScript(f'{sudo} {cmd1}') and ShellScript(f'{sudo} {cmd2}'),
(Pdb) p cmd1, cmd2
('echo subscription-manager repos --enable codeready-builder-for-rhel-8-$(arch)-rpms', 
 'echo dnf -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm')
(Pdb) self.guest.execute(ShellScript(f'{sudo} {cmd1}') and ShellScript(f'{sudo} {cmd2}'), silent=False)
CommandOutput(stdout='dnf -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm\n', stderr='')
(Pdb)   
#^^^^^^^^ CommandOutput(stdout='subscription-manager repos --enable codeready-builder-for-rhel-8-$(arch)-rpms\n', stderr='') 
#^^^^^^^^ The stdout of cmd1 was not found !!!
```

Any idea/suggestion?

Reply:
Mea culpa, can you try `&&` instead of `and`?

Reply:
`&&` doesn't work either, e.g.
```bash
$ tmt run
Traceback (most recent call last):
  File "/home/huanli/.virtualenvs/tmt/bin/tmt", line 12, in <module>
    import tmt.cli
...<snip>...
  File "/usr/lib64/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/steps/prepare/feature.py", line 110
    self.guest.execute(ShellScript(f'{sudo} {cmd1}') && ShellScript(f'{sudo} {cmd2}'),
                                                      ^
SyntaxError: invalid syntax
```

And I gave it a try to use `+`,  it works. e.g.
```python
107  ->	        if guest_distro == Distro.FEDORA:
108  	            self.info('Enable EPEL on Fedora, do nothing ...')
109  	            sudo = 'sudo' if self.guest.facts.is_superuser is False else ''
110  	            cmd1 = f'echo dnf config-manager --set-enabled {FEDORA_REPO}'
111  	            cmd2 = f'echo dnf -y install {" ".join(FEDORA_PACKAGES)}'
112  	            self.guest.execute(ShellScript(f'{sudo} {cmd1}') + ShellScript(f'{sudo} {cmd2}'), silent=False)
```

When `+` is used, the related test log looks like:
```bash
16:35:41         Enable EPEL on Fedora, do nothing ...                                                  
16:36:07         Run command: /bin/bash -c 'sudo echo dnf config-manager --set-enabled powertools; sudo echo dnf -y install epel-release epel-next-release'
16:36:07             cmd: /bin/bash -c 'sudo echo dnf config-manager --set-enabled powertools; sudo echo dnf -y install epel-release epel-next-release'
```

That is, if **`+`** is used as `ShellScript(cmd1) + ShellScript(cmd2)`, the final shell script uses **`;`** to join `cmd1` and `cmd2`.

Reply:
Once again, I'm sorry, I got lost in the code. The definitive guide :)

```python
>>> str(ShellScript('/bin/true') & ShellScript('/bin/false'))
'/bin/true && /bin/false'
>>> str(ShellScript('/bin/true') | ShellScript('/bin/false'))
'/bin/true || /bin/false'
>>> str(ShellScript('/bin/true') + ShellScript('/bin/false'))
'/bin/true; /bin/false'
>>> 
```

Reply:
Thanks very much, will use `&` or `|` if needed.

Reply:
Hi @happz, I tested the patch via `connect` for both `CentOS Stream 9` and `RHEL 8`. The results look good. e.g.

```bash
# RHEL8
...
        prepare task #1: default-0 on default-0
        how: feature
        order: 50
        Enable EPEL on RHEL 8
        Execute command 'export TMT_PLAN_DATA=/var/tmp/tmt/run-005/plans/foo/data; export TMT_TREE=/var/tmp/tmt/run-005/plans/foo/tree; subscription-manager repos --enable codeready-builder-for-rhel-8-$(arch)-rpms; dnf -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm' on guest 'kvm-03-guest06.hv2.lab.eng.bos.redhat.com'.
        Run command: sshpass -p redhat123 ssh -oForwardX11=no -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oServerAliveInterval=60 -oServerAliveCountMax=5 -oIdentitiesOnly=yes -oPasswordAuthentication=yes -S/run/user/1001/tmt/tmpvmypknpu root@kvm-03-guest06.hv2.lab.eng.bos.redhat.com 'export TMT_PLAN_DATA=/var/tmp/tmt/run-005/plans/foo/data; export TMT_TREE=/var/tmp/tmt/run-005/plans/foo/tree; subscription-manager repos --enable codeready-builder-for-rhel-8-$(arch)-rpms; dnf -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm'
        out: This system has no repositories available through subscriptions.
        out: Updating Subscription Management repositories.
        out: Unable to read consumer identity
        out: This system is not registered to Red Hat Subscription Management. You can use subscription-manager to register.
        out: beaker-AppStream-debuginfo                       25 kB/s | 1.5 kB     00:00    
...<snip>.........    
        out: beaker-tasks                                     31 kB/s | 1.3 kB     00:00    
        out: epel-release-latest-8.noarch.rpm                148 kB/s |  25 kB     00:00    
        out: Dependencies resolved.
        out: ================================================================================
        out:  Package             Architecture  Version            Repository           Size
        out: ================================================================================
        out: Installing:
        out:  epel-release        noarch        8-19.el8           @commandline         25 k
        out: 
        out: Transaction Summary
        out: ================================================================================
        out: Install  1 Package
        out: 
        out: Total size: 25 k
        out: Installed size: 35 k
        out: Downloading Packages:
        out: Running transaction check
        out: Transaction check succeeded.
        out: Running transaction test
        out: Transaction test succeeded.
        out: Running transaction
        out:   Preparing        :                                                        1/1 
        out:   Installing       : epel-release-8-19.el8.noarch                           1/1 
        out:   Running scriptlet: epel-release-8-19.el8.noarch                           1/1 
        out: Many EPEL packages require the CodeReady Builder (CRB) repository.
        out: It is recommended that you run /usr/bin/crb enable to enable the CRB repository.
        out: 
        out:   Verifying        : epel-release-8-19.el8.noarch                           1/1 
        out: Installed products updated.
        out: 
        out: Installed:
        out:   epel-release-8-19.el8.noarch                                                  
        out: 
        out: Complete!
```

Right here I have a question, what should we do for Fedora?

Reply:
No idea, probably a no-op. An info message about "nothing to do on Fedora for EPEL" & move on? @psss @thrix any idea?

Reply:
For `fedora` I'd just skip doing anything. Info message sounds good to me.